File: app/src/main/java/amagi82/modularcharactersheetcreator/models/modules/Stat.java
package amagi82.modularcharactersheetcreator.models.modules;

import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import auto.parcelgson.AutoParcelGson;

/*
    Used to fill a single row of a stat block
 */
@AutoParcelGson
public abstract class Stat implements Parcelable {
    @NonNull public abstract String category();
    @Nullable public abstract String specialty();
    public abstract int valueMin();
    public abstract int value();
    public abstract int valueTemp();
    public abstract int valueMax();
    public abstract int numStars();

    public Stat() {}

    public static Stat createDefault(String category, int valueMin, int valueMax){
        return builder().category(category).valueMin(valueMin).value(valueMin).valueTemp(valueMin).valueMax(valueMax).numStars(valueMax).build();
    }

    @AutoParcelGson.Builder
    public abstract static class Builder {
        public abstract Builder category(String category);
        public abstract Builder specialty(String specialty);
        public abstract Builder valueMin(int valueMin);
        public abstract Builder value(int value);
        public abstract Builder valueTemp(int valueTemp);
        public abstract Builder valueMax(int valueMax);
        public abstract Builder numStars(int numStars);
        public abstract Stat build();

        Builder() {}
    }

    public static Builder builder() {
        return new AutoParcelGson_Stat.Builder()
                .category("")
                .valueMax(5)
                .numStars(5);
    }

    public abstract Builder toBuilder();
}

File: app/src/main/java/amagi82/modularcharactersheetcreator/models/modules/Stat.kt
package amagi82.modularcharactersheetcreator.models.modules

import nz.bradcampbell.paperparcel.PaperParcel
import nz.bradcampbell.paperparcel.PaperParcelable

/*
    Used to fill a single row of a stat block
 */
@PaperParcel
data class Stat(val category: String = "", val specialty: String? = null, val valueMin: Int = 0, val value: Int = 0, val valueTemp: Int = 0, val valueMax: Int = 5, val numStars: Int = 5) : PaperParcelable {
    @JvmField val CREATOR = PaperParcelable.Creator(Stat::class.java)
}